mc-sema/cfgToLLVM/x86Instrs_Branches.cpp

```diff
@@ -755,181 +675,51 @@ static InstTransResult doCallPCExtern(BasicBlock *&b, std::string target,
-  }
-  // is this referencing global data?
-  else if (ip->has_mem_reference) {
-    doCallM<32>(block, ip, MEM_REFERENCE(0));
-    ret = ContinueBlock;
-    // is this a simple address computation?
-  } else {
-    doCallM<32>(block, ip, ADDR_NOREF(0));
-    ret = ContinueBlock;
-  }
-
-  return ret;
-}
-
-static InstTransResult translate_CALL32r(NativeModulePtr natM,
-                                         BasicBlock *&block, InstPtr ip,
-                                         MCInst &inst) {
-  const MCOperand &tgtOp = inst.getOperand(0);
-  //we are calling a register! this is VERY EXCITING
-  //first, we need to know which register we are calling. read that
-  //register, then make a call to the external procedure.
-  //the external procedure has a signature of
-  // void do_call_value(Value *loc, struct regs *r);
-
-  TASSERT(inst.getNumOperands() == 1, "");
-  TASSERT(tgtOp.isReg(), "");
-
-  //read the register
-  Value *fromReg = x86::R_READ<32>(block, tgtOp.getReg());
-
-  Module *M = block->getParent()->getParent();
-  const std::string &triple = M->getTargetTriple();
-  if (triple != WINDOWS_TRIPLE) {
-    x86::writeFakeReturnAddr(block);
-  }
-
-  doCallV(block, ip, fromReg);
-
-  return ContinueBlock;
-}
-
-static InstTransResult translate_JMP32r(NativeModulePtr natM,
-                                        BasicBlock *&block, InstPtr ip,
-                                        MCInst &inst) {
-  const MCOperand &tgtOp = inst.getOperand(0);
-
-  TASSERT(inst.getNumOperands() == 1, "");
-  TASSERT(tgtOp.isReg(), "");
-
-  //read the register
-  Value *fromReg = x86::R_READ<32>(block, tgtOp.getReg());
-
-  if (ip->has_jump_table()) {
-    // this is a jump table that got converted
-    // into a table in the data section
-    llvm::dbgs() << __FUNCTION__ << ": jump table via register: "
-                 << to_string<VA>(ip->get_loc(), std::hex) << "\n";
-
-    BasicBlock *defaultb = nullptr;
-
-    doJumpTableViaSwitchReg(block, ip, fromReg, defaultb, 32);
-    TASSERT(defaultb != nullptr, "Default block has to exit");
-    // fallback to doing do_call_value
-    doCallV(defaultb, ip, fromReg);
-    writeLocalsToContext(defaultb, x86::REG_SIZE, ABIRetStore);
-    ReturnInst::Create(defaultb->getContext(), defaultb);
-    return EndCFG;
-
-  } else {
-    // translate the JMP32r as a call/ret
-    llvm::dbgs() << __FUNCTION__ << ": regular jump via register: "
-                 << to_string<VA>(ip->get_loc(), std::hex) << "\n";
-    doCallV(block, ip, fromReg);
-    writeLocalsToContext(block, x86::REG_SIZE, ABIRetStore);
-    ReturnInst::Create(block->getContext(), block);
-    return EndCFG;
+  } else if (ip->has_mem_reference) {
+    doCallM<width>(block, ip, MEM_REFERENCE(0));
+    // clear stack
+    return doRet<width>(block);
   }
 }
 
-static InstTransResult translate_JMP64r(NativeModulePtr natM,
+template <int width>
+static InstTransResult translate_JMPr(NativeModulePtr natM,
```

Comment:
can you handle this by doing a force tail-call (i.e. a `jmp`) to `do_call_value`?